# Introduction
ShinyCell is an extremely usefull R package generated by [Ouyang et al. ShinyCell: Simple and sharable visualisation of single-cell gene expression data. Bioinformatics, doi:10.1093/bioinformatics/btab209](http://dx.doi.org/10.1093/bioinformatics/btab209) that allows users to simply create interactive Shiny-based web applications to visualise single-cell data. Since ShinyCell designed for using singlecell dataset objects (such as h5ad, loom, Seurat,SingleCellExperiment) thus, in this repository we share required files to make it compatible with Debit-seq and spatial maps.

# How to run 
as [here](https://github.com/SGDDNB/ShinyCell#quick-start-guide) explains, to run the app locally, first make a folder with the favorite name and copy all of the shared files (that include .rds, .h5 and .txt files) plus two important .R files server.R and ui.R (that can be found in this repository) into the new created folder as shown below.

![alt text](https://github.com/atlasxomics/ShinyCell/blob/main/images/Screenshot%202022-11-07%20at%205.52.56%20PM.png)

Next, you need to browse to the Rstudio working directory and change it to the new created folder as shown below in "R Sessions" secion.

![alt text](https://github.com/atlasxomics/ShinyCell/blob/main/images/Screenshot%202022-11-07%20at%205.56.40%20PM.png)

Then open either server.R or ui.R using RStudio and click on "Run App" in the top right corner. Now, you are all set to explore your data!

![alt text](https://github.com/atlasxomics/ShinyCell/blob/main/images/Screenshot%202022-11-07%20at%2010.28.53%20AM.png)

# Required Libraires
make sure your RStudio already installed the following librarires: shiny, shinyhelper, data.table, Matrix, DT, magrittr, ggplot2, ggrepel, hdf5r, ggdendro, gridExtra, ggseqlogo and circlize. Since all of them are CRAN packages you can install them by the following code:

install.packages(c("shiny", "shinyhelper", "data.table", "Matrix", "DT", "magrittr","ggplot2","ggrepel","hdf5r","ggdendro","gridExtra", "ggseqlogo", "circlize"))

For ploting heatmaps, ComplexHeatmap used and it needs to be installed by BiocManager as explained [here](https://www.bioconductor.org/packages/release/bioc/html/ComplexHeatmap.html). To install this package, start R use the following codes:


if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")


BiocManager::install("ComplexHeatmap")




# Quick Start Guide
The shiny app currently contains eleven tabs , and this might be developed to more tabs in future. 
In the tab called Bubbleplot/Heatmap user can plot heatmap of the input genes. This tab is using Complexheatmap package and depending on the number of imported genes and the user local system, it might take a few minutes. This tab is under development and currently only support Heatmaps. 
